### ISSUE TITLE
Stereochemistry issues with SmallMoleculeSetProposalEngine

### ISSUE THREAD

**Issue by jchodera** [choderalab/perses#120]
The errors I was seeing in #116 look to be discrepancies in the SMILES strings between the proposed molecule and what the SMILES string was intended to be:

```
======================================================================
ERROR: Testing MultiTargetDesign sampler with T4LysozymeInhibitorsTestSystem transfer free energy from vacuum -> explicit
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/choderaj/miniconda/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/Users/choderaj/github/choderalab/perses/perses.choderalab/perses/samplers/samplers.py", line 948, in run
    self.update()
  File "/Users/choderaj/github/choderalab/perses/perses.choderalab/perses/samplers/samplers.py", line 933, in update
    self.update_samplers()
  File "/Users/choderaj/github/choderalab/perses/perses.choderalab/perses/samplers/samplers.py", line 903, in update_samplers
    sampler.update()
  File "/Users/choderaj/github/choderalab/perses/perses.choderalab/perses/samplers/samplers.py", line 825, in update
    self.update_sampler()
  File "/Users/choderaj/github/choderalab/perses/perses.choderalab/perses/samplers/samplers.py", line 799, in update_sampler
    self.sampler.update()
  File "/Users/choderaj/github/choderalab/perses/perses.choderalab/perses/samplers/samplers.py", line 687, in update
    self.update_sampler()
  File "/Users/choderaj/github/choderalab/perses/perses.choderalab/perses/samplers/samplers.py", line 606, in update_sampler
    topology_proposal = self.proposal_engine.propose(system, topology)
  File "/Users/choderaj/github/choderalab/perses/perses.choderalab/perses/rjmc/topology_proposal.py", line 925, in propose
    raise Exception("smiles_new (%s) != proposed_mol_smiles (%s)" % (smiles_new, proposed_mol_smiles))
Exception: smiles_new (c1ccc(cc1)/C=N/O) != proposed_mol_smiles (ON=Cc1ccccc1)
```

I wonder if this is due to differences in how OpenEye SMILES parsing is done. The input SMILES library may not be the same flavor of canonical isomeric SMILES, so perhaps the first thing SmallMoleculeSetProposal needs to do is to check the input SMILES strings and either transform them into canonical isomeric SMILES or raise an Exception.

**Comment by pgrinaway** [comment]
Yeah, you're right. I'll do a quick fix here. I think the only way to actually make canonical isomeric SMILES would be `input_smiles` ---> `OEMol` ---> `IsoCanSMILES`, which is pretty easy.

**Comment by jchodera** [comment]
Sounds good!

I see you are addressing this in #121

**Comment by pgrinaway** [comment]
Yep. Once relevant tests pass there I will close this issue.

**Comment by jchodera** [comment]
As discussed earlier today, I'll try to sanitize all the input SMILES decks.

I'll try to put this into a utility function that allows one of several behaviors:
* raise an `Exception` if any of the input SMILES are not valid canonical isomeric SMILES
* eliminate molecules that are not valid canonical isomeric SMILES
* expand molecules that do not specify stereochemistry into all possible enantiomers/diastereomers

**Comment by jchodera** [comment]
#128 should take care of this.

**Comment by jchodera** [comment]
OK, this one is still giving me trouble:
```
['c1ccc(cc1)/C=C/C=O', 'c1ccc(cc1)C=CC=O']
```
One of these is the SMILES it is supposed to propose, and the other is the one that is actually proposed.  Any idea what the difference is?

**Comment by jchodera** [comment]
OK, it looks like this is an issue with cis- and trans- double-bonds.  

We could add stereochemistry expansion of bonds to #128, but I'm not sure the geometry engine respects cis- and trans- information in proposing new molecules.  We would need some way to ensure we properly deal with these in both the MCSS match and the geometry proposals if we want to use explicit bond stereochemistry.  The alternative would seem to be to *not* use bond stereochemistry and sample over both isomers in the simulations, but that might be suboptimal.

**Comment by jchodera** [comment]
If we wanted to take this alternative route and *not* use bond stereochemistry, there is a combination of options that would give us the appropriate level of SMILES strings.
See the OEChem documentation on [`ISOMERIC`](https://docs.eyesopen.com/toolkits/python/oechemtk/OEChemConstants/OESMILESFlag.html#OEChem::OESMILESFlag::ISOMERIC). We would omit the `BondStereo` flavor.

**Comment by jchodera** [comment]
My thinking so far is that there are two simple ways to address the cis/trans double bond issue and the ring formation yield issue:
* **Increase the yield by adding torsions.** Add some extra periodicity=1 torsions that are intended to preserve bond stereochemistry and ring integrity.  `SmallMoleculeProposalEngine` could identify these bonds and their phase and list them in `metadata`, and we could pass them on to `SystemGenerator` to add extra torsions that preserve the stereochemistry during proposals.
* **Provide a way for `GeometryEngine` to reject incorrect configurations.** If the resulting bond(s) have the wrong stereochemistry or rings don't close, we can have `GeometryEngine` reject the proposal and potentially keep trying to propose new molecules until something works.  Yields would likely be reduced as the molecules have more stereo-bonds or rings, however.

**Comment by jchodera** [comment]
Even *with* my `sanitizeSMILES` function to drop or enumerate stereochemistry centers, I still see errors like this that don't appear to be due to the bond stereochemistry issue:
```
================================================================================
SAMS sampler iteration     1
--------------------------------------------------------------------------------
Expanded Ensemble sampler iteration        1
................................................................................
MCMC sampler iteration 1
Taking 2500 steps of Langevin dynamics...
Final energy is     -438.081 kT
................................................................................
Updating chemical state with ncmc-geometry-ncmc scheme...
Proposing new topology...
proposed SMILES string: CC[C@@H](c1nc2cccc(c2c(=O)n1c3ccccc3)F)Nc4c-5ncnc5nc[nH]4
LOG: MOL
LOG: MOL
Traceback (most recent call last):
  File "perses/tests/testsystems.py", line 792, in <module>
    run_kinase_inhibitors()
  File "perses/tests/testsystems.py", line 789, in run_kinase_inhibitors
    testsystem.sams_samplers[environment].run(niterations=1000)
  File "/Users/choderaj/miniconda/lib/python2.7/site-packages/perses-0.1-py2.7.egg/perses/samplers/samplers.py", line 922, in run
    self.update()
  File "/Users/choderaj/miniconda/lib/python2.7/site-packages/perses-0.1-py2.7.egg/perses/samplers/samplers.py", line 906, in update
    self.update_sampler()
  File "/Users/choderaj/miniconda/lib/python2.7/site-packages/perses-0.1-py2.7.egg/perses/samplers/samplers.py", line 874, in update_sampler
    self.sampler.update()
  File "/Users/choderaj/miniconda/lib/python2.7/site-packages/perses-0.1-py2.7.egg/perses/samplers/samplers.py", line 752, in update
    self.update_state()
  File "/Users/choderaj/miniconda/lib/python2.7/site-packages/perses-0.1-py2.7.egg/perses/samplers/samplers.py", line 650, in update_state
    topology_proposal = self.proposal_engine.propose(system, topology)
  File "/Users/choderaj/miniconda/lib/python2.7/site-packages/perses-0.1-py2.7.egg/perses/rjmc/topology_proposal.py", line 939, in propose
    raise Exception("smiles_new (%s) != proposed_mol_smiles (%s)" % (smiles_new, proposed_mol_smiles))
Exception: smiles_new (CCC(c1nc2cccc(c2c(=O)n1c3ccccc3)F)Nc4c-5ncnc5nc[nH]4) != proposed_mol_smiles (CC[C@@H](c1nc2cccc(c2c(=O)n1c3ccccc3)F)Nc4c-5ncnc5nc[nH]4)
```
Anyone know what might be going on here?  It looks like there is a stereocenter at the third `C` that is expanded here to `C@@H` in this case, but somehow this isn't flagged as an undefined stereocenter in [`sanitizeSMILES()`](https://github.com/choderalab/perses/blob/master/perses/tests/utils.py#L348-L419).  Any ideas?

**Comment by jchodera** [comment]
I've figured out why `generateOEMolFromTopologyResidue` was generating stereochemistry: By default, it was populating the configuration with a random geometry from `OEOmega`.  This behavior is now off by default as of https://github.com/choderalab/openmoltools/pull/190

Updating `openmoltools` should fix this.  That reminds me also that I need to update `openmoltools-dev` to be built properly for multiple platforms...

**Comment by jchodera** [comment]
I believe this is fixed by #639
